### ISSUE TITLE
Unresolved citations when building pdf-docs

### ISSUE THREAD

**Issue by strogdon** [cschwan/sage-on-gentoo#543]
On Gentoo I see things like
```
Output written on tensor_free_modules.pdf (205 pages, 693136 bytes).
Transcript written on tensor_free_modules.log.
Latexmk: Index file 'tensor_free_modules.idx' was written
Latexmk: Citation `../references/index:god1968' on page 3 undefined
Latexmk: Citation `../references/index:god1968' on page 3 undefined
Latexmk: Citation `../references/index:lan2002' on page 3 undefined
Latexmk: Citation `../references/index:lan2002' on page 3 undefined
Latexmk: Citation `../references/index:god1968' on page 37 undefined
Latexmk: Citation `../references/index:god1968' on page 37 undefined
Latexmk: Citation `../references/index:lee2013' on page 37 undefined
Latexmk: Citation `../references/index:lee2013' on page 37 undefined
Latexmk: Citation `../references/index:one1983' on page 37 undefined
Latexmk: Citation `../references/index:one1983' on page 37 undefined
Latexmk: Citation `../references/index:god1968' on page 41 undefined
Latexmk: Citation `../references/index:god1968' on page 41 undefined
Latexmk: Citation `../references/index:lan2002' on page 41 undefined
Latexmk: Citation `../references/index:lan2002' on page 41 undefined
Latexmk: Citation `../references/index:con2015' on page 49 undefined
Latexmk: Citation `../references/index:con2015' on page 49 undefined
Latexmk: Citation `../references/index:god1968' on page 49 undefined
Latexmk: Citation `../references/index:god1968' on page 49 undefined
Latexmk: Citation `../references/index:lan2002' on page 49 undefined
Latexmk: Citation `../references/index:lan2002' on page 49 undefined
Latexmk: Citation `../references/index:god1968' on page 55 undefined
Latexmk: Citation `../references/index:god1968' on page 55 undefined
Latexmk: Citation `../references/index:lee2013' on page 56 undefined
Latexmk: Citation `../references/index:lee2013' on page 56 undefined
Latexmk: Citation `../references/index:one1983' on page 56 undefined
Latexmk: Citation `../references/index:one1983' on page 56 undefined
Latexmk: Citation `../references/index:con2013' on page 89 undefined
Latexmk: Citation `../references/index:con2013' on page 89 undefined
Latexmk: Citation `../references/index:lan2002' on page 89 undefined
Latexmk: Citation `../references/index:lan2002' on page 89 undefined
Latexmk: Citation `../references/index:god1968' on page 96 undefined
Latexmk: Citation `../references/index:god1968' on page 96 undefined
Latexmk: Citation `../references/index:lan2002' on page 96 undefined
Latexmk: Citation `../references/index:lan2002' on page 96 undefined
Latexmk: Citation `../references/index:god1968' on page 105 undefined
Latexmk: Citation `../references/index:god1968' on page 105 undefined
Latexmk: Citation `../references/index:lan2002' on page 105 undefined
Latexmk: Citation `../references/index:lan2002' on page 105 undefined
Latexmk: Citation `../references/index:god1968' on page 117 undefined
Latexmk: Citation `../references/index:god1968' on page 117 undefined
Latexmk: Citation `../references/index:lan2002' on page 117 undefined
Latexmk: Citation `../references/index:lan2002' on page 117 undefined
Latexmk: Citation `../references/index:god1968' on page 121 undefined
Latexmk: Citation `../references/index:god1968' on page 121 undefined
Latexmk: Citation `../references/index:lan2002' on page 121 undefined
Latexmk: Citation `../references/index:lan2002' on page 121 undefined
Latexmk: Citation `../references/index:god1968' on page 127 undefined
Latexmk: Citation `../references/index:god1968' on page 127 undefined
Latexmk: Citation `../references/index:god1968' on page 132 undefined
Latexmk: Citation `../references/index:god1968' on page 132 undefined
Latexmk: Log file says output to 'tensor_free_modules.pdf'
Latexmk: Summary of warnings:
  Latex failed to resolve 50 citation(s)
Latexmk: All targets (tensor_free_modules.pdf) are up-to-date
```
and there are missing citations in the pdf file. The above references are to R. Godement and S. Lang. Perhaps some `texlive` component is missing or not up-to-date.

**Comment by kiwifb** [comment]
I haven't built the latest rc but that looks like material for sage-devel and quite likely a ticket. pdf docs building is not tested anywhere nearly as often as the html one. Is this with python 3.6 or 2.7?

**Comment by strogdon** [comment]
python 2.7. I usually don't build the pdfs on the box where vanilla is installed because of computing resources. I have a suspicion that `texlive-2019` may be needed.

**Comment by kiwifb** [comment]
That's a definite possibility.

**Comment by kiwifb** [comment]
I am going to check vanilla here.

**Comment by strogdon** [comment]
I was able to generate the pdf-docs for vanilla. I see the same missing citations.

**Comment by kiwifb** [comment]
Same here. It is definitely sage-devel material.

**Comment by strogdon** [comment]
Apparently, when both python2 and python3 are enabled python3 is used to build the pdf-docs. The missing citations then slightly different. Mainly the number of a missing citations is different.

**Comment by kiwifb** [comment]
Considering the html doc is fine, I am wondering if this is an undetected fall out of the latest sphinx update.

**Comment by kiwifb** [comment]
Still happening with texlive 2019.

**Comment by strogdon** [comment]
There seems to be something structurally incorrect with how the citations are implemented. From `tensor_free_modules.tex` there is
```
REFERENCES:
\begin{itemize}
\item {}
Chap. 10 of R. Godement : \sphinxstyleemphasis{Algebra} \sphinxcite{../references/index:god1968}

\item {}
Chap. 3 of S. Lang : \sphinxstyleemphasis{Algebra} \sphinxcite{../references/index:lan2002}

\end{itemize}
```
I don't see how this can possibly work. The `\sphinxcite` is the equivalent if the Latex `\cite`. To get the citations to work there should be a `\begin{thebibliography} ... \end{thebibliography}` environment in the .tex file that has`\bibitem` entries, and I don't see the environment. Furthermore, the citations are being referenced to a folder where the `\bibitem{}` that gives the `\cite{}` meaning is contained in a `.tex` file, i.e. `references/references.tex`. I've never seen this type of implementation.

**Comment by kiwifb** [comment]
Because this is clearly an upstream problem I think discussion should be taking place on sage-devel rather than here.

**Comment by strogdon** [comment]
You're right. Perhaps a ticket should be opened so comments can be organized. We're not going to fix it here.

**Comment by kiwifb** [comment]
And this is now https://trac.sagemath.org/ticket/28059

**Comment by strogdon** [comment]
The proposed fix at https://trac.sagemath.org/ticket/28059 does not work here. I want to make sure this is not a s-o-g issue. With the indicated patch I see ALL of the unresolved Citations that I had before applying the patch. If I enter a folder where the LaTeX codes resides that has citation issues (after `ebuild sage-9999 clean compile`) and type `make` then the indicated PDF is generated and ALL Citations are properly resolved?

**Comment by kiwifb** [comment]
Do you have a patch to apply I can download somewhere?

**Comment by strogdon** [comment]
I forgot that.
```
--- a/doc/en/reference/conf_sub.py
+++ b/doc/en/reference/conf_sub.py
@@ -63,6 +63,18 @@ latex_documents = [
 ('index', name + '.tex', project, u'The Sage Development Team', 'manual')
 ]
 
+latex_elements['hyperref'] = r"""
+\usepackage{xcite}
+\usepackage{xr-hyper}
+\externaldocument[../references/]{../references/references}
+\externalcitedocument[../references/]{../references/references}
+% Include hyperref last.
+\usepackage{hyperref}
+% Fix anchor placement for figures with captions.
+\usepackage{hypcap}% it must be loaded after hyperref.
+% Set up styles of URL: it should be placed after hyperref.
+\urlstyle{same}"""
+
 #Ignore all .rst in the _sage subdirectory
 exclude_patterns = exclude_patterns + ['_sage']

```

**Comment by kiwifb** [comment]
Did I understand that it needed to be applied to all `conf_sub.py` files present?

**Comment by strogdon** [comment]
For that I'm not sure. But I only checked what was going on under `en/reference`.

**Comment by kiwifb** [comment]
I can't realistically do something before Monday (and even then, I may be to busy with RL work next week).

**Comment by strogdon** [comment]
No rush but I believe this is the problem. From `tensor_free_modules.log`
```
(/usr/share/texmf-dist/tex/latex/xcite/xcite.sty
Package: xcite 2011/09/02 v1.0 eXternal Citations (EG)
)
(/usr/share/texmf-dist/tex/latex/hyperref/xr-hyper.sty
Package: xr-hyper 2000/03/22 v6.00beta4 eXternal References (DPC)
)

Package xr Warning:
No file ../references/references.aux
LABELS NOT IMPORTED.
 on input line 35.

Package xc Warning:
No file ../references/references.aux
LABELS NOT IMPORTED.
 on input line 35.
```
For the patch to work the stuff under `../references/*` has to be built (i.e. LaTeXed). And I don't think there is any way to insure that it will be, especially if things are built in parallel. So I think it's possible to appear on vanilla sage.

**Comment by strogdon** [comment]
As noted above a partial fix has been merged into vanilla and should appear in `8.9.rc1`. Since the vanilla people are aware of the issue any further tweaks should appear on [Trac](https://trac.sagemath.org).
